src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -23913,26 +23949,136 @@ static void M1<T>(T t1)
+                );
+        }
+
+        [Fact, WorkItem(36009, "https://github.com/dotnet/roslyn/issues/36009")]
+        public void DisallowNull_Parameter_NullableValueType()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    static void F5<T>([DisallowNull]T? t) where T : struct { }
+    static void M5<T>(T? t5) where T : struct
+    {
+        F5(t5); // 1
+        if (t5 == null) return;
+        F5(t5);
+    }
+}";
+            var comp = CreateNullableCompilation(new[] { DisallowNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (7,12): warning CS8607: A possible null value may not be passed to a target marked with the [DisallowNull] attribute
+                //         F5(t5);
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "t5").WithLocation(7, 12)
+                );
+        }
+
+        [Fact, WorkItem(36009, "https://github.com/dotnet/roslyn/issues/36009")]
+        public void DisallowNull_RefParameter_NullableValueType()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    static void F5<T>([DisallowNull]ref T? t) where T : struct { }
+    static void M5<T>(T? t5) where T : struct
+    {
+        F5(ref t5); // 1
+        if (t5 == null) return;
+        F5(ref t5);
+    }
+}";
+            var comp = CreateNullableCompilation(new[] { DisallowNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (7,16): warning CS8607: A possible null value may not be passed to a target marked with the [DisallowNull] attribute
+                //         F5(ref t5); // 1
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "t5").WithLocation(7, 16)
+                );
+        }
+
+        [Fact, WorkItem(36009, "https://github.com/dotnet/roslyn/issues/36009")]
+        public void DisallowNull_InParameter_NullableValueType()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    static void F5<T>([DisallowNull]in T? t) where T : struct { }
+    static void M5<T>(T? t5) where T : struct
+    {
+        F5(in t5); // 1
+        if (t5 == null) return;
+        F5(in t5);
+    }
+}";
+            var comp = CreateNullableCompilation(new[] { DisallowNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (7,15): warning CS8607: A possible null value may not be passed to a target marked with the [DisallowNull] attribute
+                //         F5(in t5); // 1
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "t5").WithLocation(7, 15)
+                );
+        }
+
+        [Fact, WorkItem(36009, "https://github.com/dotnet/roslyn/issues/36009")]
+        public void DisallowNull_ByValParameter_NullableValueTypeViaConstraint()
```

Comment:
>DisallowNull_ByValParameter_NullableValueTypeViaConstraint [](start = 20, length = 58)

I am not sure what are we trying to test here, struct constraint doesn't allow nullable value types and U is a nullable value type from constraints.